### DISCUSSION TITLE
Current status W.R.T. react-native-sqlite-storage

### DISCUSSION THREAD

**Discussion by MacKenzieHnC** [microsoft/react-native-windows#11048] [Q&A] [answered]
tl;dr: Is `react-native-sqlite-storage` 
-known to work with current RNW?
-known not to work with current RNW?
-or unknown if it works with current RNW?

I understand this is an external module, and I am not asking anyone to solve this for me.

I am about to embark on fully understanding and either fixing it or writing some more thorough documentation for it (whichever is necessary), and I just want to know before I do if anyone here knows of any reason it shouldn't be working. Something like [this](https://github.com/microsoft/react-native-windows/issues/6033) where some change in this repo would have obviously broken functionality there.

Through a lot of digging, I've finally got things to a point where I feel like it should be working, but I'm going to have to dig more through its code to figure out why it isn't.

So is there any obvious reason from this repo why `react-native-sqlite-storage` isn't working for RNW anymore? Or does anyone know for a fact that it _does_ work and that this will end up being user error?

EDIT (random complaint): Seriously, what is the deal with no one following up after they ask for help with this module?? Every single issue dead-ends with "needs author feedback". Infuriating.

**Comment by MacKenzieHnC** [comment, marked as answer]
Okay, I think I have answered my own question and solved the issue. If possible, could an expert in this language please confirm that `.GetResults()` is misused in the orignal code [here](https://github.com/andpor/react-native-sqlite-storage/blob/b6aeac439b9d038bbb9d2e16812d512c38246775/platforms/windows/SQLitePlugin/SQLitePlugin.cpp#L310).

When I was running it, I get a `method called at an unexpected time` error. Deleting `getResults()` fixes the issue and reinstates full functionality. For more detailed instructions, [these are the steps I took to get pre-populated databases working with RNW](https://github.com/andpor/react-native-sqlite-storage/issues/491#issuecomment-1368515615).

**Reply by MacKenzieHnC** [reply]
Of *course* it wasn't that simple. I've narrowed down what the issue is, but this language is baffling me.

The operation attempts to copy a file and throws an exception if the file already exists.

But it seems like all the available versions of trying to await that result cause a fatal exception if it does throw. It's like it still tries to call `.get()` even if the exception is caught.

All it's supposed to be is a synchronous file copy. Is there really no way to do that? There has to be, right?

**Reply by MacKenzieHnC** [reply]
Okay, I (finally?) think it was that easy. I just failed to delete the `onFailure()` method out of the `catch` block so as to fully ignore the exception. Would still like someone to glance at those 2 lines I changed, but I feel pretty good about it.

**Reply by chrisglein** [reply]
> Would still like someone to glance at those 2 lines I changed, but I feel pretty good about it.

Do you have a PR that contains your edits that we can look at? It'd help to look at it all in context. Thanks for looking into this and trying to get it to work.

To answer your original question, we hadn't heard of any reports of the module not working. Although the work to enable it was done a couple years ago and stuff happens version to version.

**Reply by MacKenzieHnC** [reply]
Thank you for your response. [This](https://github.com/andpor/react-native-sqlite-storage/pull/550) is my PR on that repo (though it looks to me like the repo is essentially abandoned, so I don't expect it to get accepted). There have been a few complaints here, but primarily they were on that repo. NOTE: In the original code, that `CopyDbAsync` call ends in a `.GetResults()`. I had erroneously thought I could just delete `.GetResults()` to get the desired effect, so the git change looks like that call wasn't there.

The big nightmare I ran into as a full `react-native-windows`/UWP/VS newbie was just figuring out how to include files. Once I had that it was a lot easier to know what was broken and find how to fix it.

If I'm understanding the effect of `FailIfExists` coupled with the original throw of `onFailure`, then the original code was refusing to open the db if it already existed, which sounds ridiculous like I have to be wrong, but it seems like that was the case, and now it functions fine after my hack. Also `.GetResult` was firing out of sync with `CopyAsync` anyway, causing a catastrophic exception. So it seems that either I'm very wrong about what was happening or everyone else just gave up eventually.

**Comment by MacKenzieHnC** [comment]
Related question: what's the difference between storing data in `ms-appx` vs `C:\Users\...\Packages\<PackageCode>\whatever`? This code does some work to copy the database to the location the app's running from and I can't figure out why. Would that allow you to store it on a server or something in a way that `ms-appx` wouldn't? Or is it safe to simplify it to just use `ms-appx`?

**Comment by chrisglein** [comment]
Attempting to answer this question:
> EDIT (random complaint): Seriously, what is the deal with no one following up after they ask for help with this module?? Every single issue dead-ends with "needs author feedback". Infuriating.

I can attempt to fill in the story there. When it became possible to add Windows support for community modules there was a need to track which were the most needed modules that were blocking apps. There's over a thousand of modules [registered in the directory](https://reactnative.directory/), but only a few dozen are _really_ needed for the majority of apps. So we opened up the ability for people to file requests on this repo to try to learn about which ones were most needed (even though the source and general support for those modules lives in other repos). A small handful of folks across a couple different companies then spent some time trying to bridge those gaps and implement some modules. At some point a threshold was crossed of "good enough coverage" that it didn't make sense to keep pushing in that way. That's why we no longer have that form on this repo - really the way to get that done is to engage on the repo of the community module owner themselves. We haven't closed the remaining issues on this repo because technically those modules _do_ still need Windows implementations, but it's not an area of active investment (and if it were, it'd probably have to wait for Fabric anyway).

As for why they have "needs author feedback", that's supposed to be applied when there's an active question that the issue author should follow up on. Often that's missing parts of the template, such as details on how the module is being used that would contextualize its relative priority. But it shouldn't be on most of them; that sounds unintentional.

**Reply by MacKenzieHnC** [reply]
Okay, that makes a lot of sense, and explains it on this repo, although it still feels rude to me for people to not follow up after they ask a question; I get people are busy, but for me, a big part of the ethos of being a programmer is solving problems so the next person doesn't have to, so it's always frustrating when people just give up and I still have to start from scratch.

But it's also true on the other repo. Everyone who got stuck anywhere near this problem just stopped responding eventually. Or didn't bother digging in to find the root of it, so I ended up spending >100 hours in an extremely arcane language I don't speak trying to find it. And that's true for both the problem of including arbitrary files as well as this issue.

I'm praying you'll confirm my workaround and this will just finally be solved.
